File: tools/sentinel/src/lib/tools/clipper.ts
/**
 * ✂️ Clipper — Fetch a webpage and save it as Markdown to vault
 *
 * Fetch strategy:
 *   1. Jina AI (r.jina.ai) — fast, handles most pages
 *   2. Local readability fallback via fetch + @mozilla/readability — works offline
 *
 * Usage (direct): tsx src/lib/tools/clipper.ts <url> [target_dir]
 */

import { writeFile, mkdir } from 'fs/promises';
import { existsSync } from 'fs';
import { join } from 'path';
import { config } from 'dotenv';

config();

function sanitizeFilename(title: string): string {
    return title.replace(/[\\/*?:<>|"]/g, '').trim().substring(0, 100);
}

/**
 * Strategy 1: Use Jina AI to convert a URL to Markdown.
 */
async function fetchViaJina(url: string): Promise<{ title: string; content: string }> {
    const jinaUrl = `https://r.jina.ai/${url}`;
    const res = await fetch(jinaUrl, {
        headers: { 'User-Agent': 'Mozilla/5.0', 'Accept': 'text/markdown' },
        signal: AbortSignal.timeout(30_000),
    });

    if (!res.ok) throw new Error(`Jina API returned ${res.status}`);
    const content = await res.text();

    let title = 'Untitled Clipper';
    for (const line of content.split('\n').slice(0, 10)) {
        if (line.startsWith('Title: ')) { title = line.slice(7).trim(); break; }
        if (line.startsWith('# ')) { title = line.slice(2).trim(); break; }
    }

    return { title, content };
}

/**
 * Strategy 2: Local fetch + lightweight HTML-to-Markdown conversion.
 * Used when Jina is unavailable or times out.
 */
async function fetchViaLocal(url: string): Promise<{ title: string; content: string }> {
    const res = await fetch(url, {
        headers: { 'User-Agent': 'Mozilla/5.0 (compatible; NeoAgent-Clipper/1.0)' },
        signal: AbortSignal.timeout(20_000),
    });

    if (!res.ok) throw new Error(`HTTP ${res.status} from ${url}`);

    const html = await res.text();

    // Extract title
    const titleMatch = html.match(/<title[^>]*>([^<]+)<\/title>/i);
    const title = titleMatch ? titleMatch[1].trim() : 'Untitled';

    // Strip scripts, styles, nav, footer for cleaner content
    const stripped = html
        .replace(/<script[\s\S]*?<\/script>/gi, '')
        .replace(/<style[\s\S]*?<\/style>/gi, '')
        .replace(/<nav[\s\S]*?<\/nav>/gi, '')
        .replace(/<header[\s\S]*?<\/header>/gi, '')
        .replace(/<footer[\s\S]*?<\/footer>/gi, '');

    // Naive HTML → Markdown conversion (handles common elements)
    const content = stripped
        .replace(/<h([1-6])[^>]*>([\s\S]*?)<\/h\1>/gi, (_, n, t) => `${'#'.repeat(+n)} ${t.replace(/<[^>]+>/g, '').trim()}\n\n`)
        .replace(/<a[^>]*href="([^"]*)"[^>]*>([\s\S]*?)<\/a>/gi, (_, href, text) => `[${text.replace(/<[^>]+>/g, '').trim()}](${href})`)
        .replace(/<strong[^>]*>([\s\S]*?)<\/strong>/gi, '**$1**')
        .replace(/<em[^>]*>([\s\S]*?)<\/em>/gi, '*$1*')
        .replace(/<li[^>]*>([\s\S]*?)<\/li>/gi, (_, t) => `- ${t.replace(/<[^>]+>/g, '').trim()}\n`)
        .replace(/<p[^>]*>([\s\S]*?)<\/p>/gi, (_, t) => `${t.replace(/<[^>]+>/g, '').trim()}\n\n`)
        .replace(/<br\s*\/?>/gi, '\n')
        .replace(/<[^>]+>/g, '')            // strip remaining tags
        .replace(/&nbsp;/g, ' ')
        .replace(/&amp;/g, '&')
        .replace(/&lt;/g, '<')
        .replace(/&gt;/g, '>')
        .replace(/&quot;/g, '"')
        .replace(/\n{3,}/g, '\n\n')
        .trim();

    return { title, content };
}

/**
 * Fetch a URL as Markdown, trying Jina first then local fallback.
 */
async function fetchMarkdown(url: string): Promise<{ title: string; content: string; source: string }> {
    try {
        const result = await fetchViaJina(url);
        return { ...result, source: 'jina' };
    } catch (jinaErr) {
        console.warn(`[Clipper] ⚠️  Jina failed (${jinaErr}), trying local fallback...`);
        try {
            const result = await fetchViaLocal(url);
            return { ...result, source: 'local' };
        } catch (localErr) {
            throw new Error(`Both strategies failed.\n  Jina: ${jinaErr}\n  Local: ${localErr}`);
        }
    }
}

export async function clipUrl(url: string, targetDir?: string): Promise<string> {
    const vaultDir = process.env.GEMINI_WORK_DIR;

    let saveDir: string;
    if (targetDir && existsSync(targetDir)) {
        saveDir = targetDir;
    } else if (targetDir && vaultDir) {
        saveDir = join(vaultDir, targetDir);
    } else if (vaultDir) {
        saveDir = join(vaultDir, '00_收集');
    } else {
        throw new Error('GEMINI_WORK_DIR not set and no absolute targetDir provided');
    }

    await mkdir(saveDir, { recursive: true });

    const { title, content, source } = await fetchMarkdown(url);
    const safeTitle = sanitizeFilename(title);
    const today = new Date().toISOString().split('T')[0];

    const markdown = `---
title: ${title}
url: ${url}
date: ${today}
type: clipper
source: ${source}
tags: [inbox]
---

${content}`;

    let filePath = join(saveDir, `${safeTitle}.md`);
    let counter = 1;
    while (existsSync(filePath)) {
        filePath = join(saveDir, `${safeTitle}_${counter++}.md`);
    }

    await writeFile(filePath, markdown, 'utf-8');
    return filePath;
}

// Run directly
if (import.meta.url === `file://${process.argv[1]}`) {
    const url = process.argv[2];
    if (!url) { console.error('Usage: tsx clipper.ts <url> [target_dir]'); process.exit(1); }
    console.log(`✂️  Clipping: ${url}`);
    const path = await clipUrl(url, process.argv[3]);
    console.log(`✅ Saved to: ${path}`);
}

File: neo-tools/refinery/clipper.py
import os
import sys
import httpx
from bs4 import BeautifulSoup
from markdownify import markdownify as md
from pathlib import Path
from datetime import datetime
import re

def sanitize_filename(title: str) -> str:
    import re
    cleaned = re.sub(r'[\\/*?:<>|"]', '', title)
    return cleaned.strip()[:100]

def fetch_via_jina(url: str) -> dict[str, str]:
    jina_url = f"https://r.jina.ai/{url}"
    headers = {'User-Agent': 'Mozilla/5.0', 'Accept': 'text/markdown'}
    with httpx.Client(timeout=30.0) as client:
        res = client.get(jina_url, headers=headers)
        res.raise_for_status()
        content = res.text
        
        title = 'Untitled Clipper'
        for line in content.split('\n')[:10]:
            if line.startswith('Title: '):
                title = line[7:].strip()
                break
            if line.startswith('# '):
                title = line[2:].strip()
                break
                
        return {"title": title, "content": content}

def fetch_via_local(url: str) -> dict[str, str]:
    headers = {'User-Agent': 'Mozilla/5.0 (compatible; NeoAgent-Clipper/1.0)'}
    with httpx.Client(timeout=20.0) as client:
        res = client.get(url, headers=headers)
        res.raise_for_status()
        html = res.text
        
        soup = BeautifulSoup(html, 'html.parser')
        title = soup.title.string.strip() if soup.title else 'Untitled'
        
        for element in soup(["script", "style", "nav", "header", "footer"]):
            element.decompose()
            
        content = md(str(soup), heading_style="ATX", default_title=True).strip()
        # Clean excessive newlines
        content = re.sub(r'\n{3,}', '\n\n', content)
        
        return {"title": title, "content": content}

def fetch_markdown(url: str) -> dict[str, str]:
    try:
        result = fetch_via_jina(url)
        result["source"] = "jina"
        return result
    except Exception as jina_err:
        print(f"[Clipper] ⚠️ Jina failed ({jina_err}), trying local fallback...", file=sys.stderr)
        try:
            result = fetch_via_local(url)
            result["source"] = "local"
            return result
        except Exception as local_err:
            raise RuntimeError(f"Both strategies failed.\n  Jina: {jina_err}\n  Local: {local_err}")

def clip_url(url: str, target_dir: str | None = None) -> str:
    from dotenv import load_dotenv
    load_dotenv()
    
    vault_dir = os.environ.get("GEMINI_WORK_DIR")
    
    if target_dir and Path(target_dir).exists():
        save_dir = Path(target_dir)
    elif target_dir and vault_dir:
        save_dir = Path(vault_dir) / target_dir
    elif vault_dir:
        save_dir = Path(vault_dir) / 'inbox'
    else:
        raise RuntimeError("GEMINI_WORK_DIR not set and no absolute target_dir provided")
        
    save_dir.mkdir(parents=True, exist_ok=True)
    
    result = fetch_markdown(url)
    title = result["title"]
    content = result["content"]
    source = result["source"]
    
    safe_title = sanitize_filename(title)
    today = datetime.now().strftime("%Y-%m-%d")
    
    markdown = f"""---
title: {title}
url: {url}
date: {today}
type: clipper
source: {source}
tags: [inbox]
---

{content}"""

    file_path = save_dir / f"{safe_title}.md"
    counter = 1
    while file_path.exists():
        file_path = save_dir / f"{safe_title}_{counter}.md"
        counter += 1
        
    file_path.write_text(markdown, encoding='utf-8')
    return str(file_path)

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python clipper.py <url> [target_dir]", file=sys.stderr)
        sys.exit(1)
        
    url = sys.argv[1]
    target_dir = sys.argv[2] if len(sys.argv) > 2 else None
    
    print(f"✂️ Clipping: {url}")
    try:
        path = clip_url(url, target_dir)
        print(f"✅ Saved to: {path}")
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)
